Parsing

drets opened the pull request:
@neongreen could you please review?

drets committed:
Initial commit.

Change parseNumber to support the Scheme standard for different bases. You may find the readOct and readHex functions useful.

drets committed:
Parse character literals.

Add a Character constructor to LispVal, and create a parser for character literals as described in R5RS.

drets committed:
Add data types, parsers to support the full numeric tower of Scheme.

https://en.wikibooks.org/wiki/Write_Yourself_a_Scheme_in_48_Hours/Parsing

Add data types and parsers to support the full numeric tower of Scheme
numeric types. Haskell has built-in types to represent many of these; check
the Prelude. For the others, you can define compound types that represent
eg. a Rational as a numerator and denominator, or a Complex as a real and
imaginary part (each itself a Real).

drets committed:
Minor.

drets committed:
Add code from the tutor; add unit tests.

neongreen commented on Main.hs:
```diff
@@ -0,0 +1,149 @@
+parseString :: Parser LispValue
+parseString = do
+  char '"'
+  x <- many1 ((char '\\' >> char '"')
```
This is wrong – if you have e.g. `\r` in your string, it'll return `r`, but it should return `'\r'` instead.

``` haskell
x <- many1 $ choice [
  string "\\\"" $> '"',
  string "\\r" $> '\r',
  string "\\\\" $> '\\',
  ...
  ]
```

neongreen replied:
Or perhaps

``` haskell
let escape c = "\\" ++ [c]
x <- many1 $ choice [
  escape '"' $> '"',
  escape 'r' $> '\r',
  escape '\\' $> '\\',
  ...
  ]
```

neongreen commented on Main.hs:
```diff
@@ -0,0 +1,149 @@
+
+parseNumberWithRadix :: Parser LispValue
+parseNumberWithRadix = do
+  t <- char 'o' <|> char 'x' <|> char 'd' <|> char 'b'
```
oneOf

neongreen commented on Main.hs:
```diff
@@ -0,0 +1,149 @@
+parseNumberWithRadix = do
+  t <- char 'o' <|> char 'x' <|> char 'd' <|> char 'b'
+  case t of
+    'o' -> parseOctal
```
This can be rewritten with `choice` and then you wouldn't need the line above

```
choice [
  char 'o' >> parseOctal, 
 ...

neongreen commented on Main.hs:
```diff
@@ -0,0 +1,149 @@
+         <|> parseString
+         <|> parseAtom
+
+parseCharacter :: Parser LispValue
```
What does this do?

drets replied:
http://www.schemers.org/Documents/Standards/R5RS/HTML/r5rs-Z-H-9.html#%_sec_6.3.4

neongreen commented on Main.hs:
```diff
@@ -0,0 +1,149 @@
+parseComplex :: Parser LispValue
+parseComplex = do
+  RealNumber x <- parseReal
+  char '+'
```
I thought complex numbers could also have the form of `a-bi`

neongreen commented on MainTest.hs:
```diff
@@ -0,0 +1,76 @@
+
+import Test.HUnit
+import Control.Monad
+import Text.ParserCombinators.Parsec hiding (spaces, try)
```
This is deprecated, import Text.Parsec instead

drets committed:
Code review fixes + parsing part 2.

Did all exercises for
https://en.wikibooks.org/wiki/Write_Yourself_a_Scheme_in_48_Hours/Parsing
except last one which I am going to skip:

Instead of using the try combinator, left-factor the grammar so that the common subsequence is its own parser. You should end up with a parser that matches a string of expressions, and one that matches either nothing or a dot and a single expression. Combining the return values of these into either a List or a DottedList is left as a (somewhat tricky) exercise for the reader: you may want to break it out into another helper function.

neongreen commented on Main.hs:
```diff
@@ -0,0 +1,199 @@
+               deriving (Eq, Show)
+
+data LispComplex =
+  LispComplex LispReal Char LispReal
```
It might be better to use `Bool` instead of `Char` or even create a special `data Sign = Positive | Negative`

drets committed:
More types.

drets committed:
Handle -10.
